### PR TITLE
Completed assignment 4

### PR DESCRIPTION
**Pull request by ZStuefen**

### CODE REVIEW

**Comment by jamesgreiner** [comment]
I will review this.

**Review by jamesgreiner — COMMENTED** [review]
# Assignment
- Create a UserController in the API projects
  - Implement endpoints for:
    - Retrieving all users ✔
    - Retriving a single user by id ✔
    - Creating a user ✔
    - Updating a user ✔
    - Deleting a user ✔
  - The UserController should function. ✔
  - Include a `.rest` file for interacting with the controller ([REST Client](https://marketplace.visualstudio.com/items?itemName=humao.rest-client)). ✔
- The `IUserRepository` interface is provided for you. You will need to provide an implementation of this interface within the SecretSanta.Business project. This implementation can simply store the user within a list, dictionary, array, or similar. ✔
- All test projects should be within a test directory ✔
- Create a unit test project for the SecretSanta.Business project. ❌
  - Unit test all public methods ❌
  - Ensure the project is named with a ".Tests" suffix ❌
  - Ensure test class is named appropriately for the class being tested: ❌
  - Name all tests in the form `<MethodBeingTested>_<ConditionBeingTested>_<ExpectedOutcome>()`. You can add  `<NoWarn>CA1707;INTL0003</NoWarn>` to remove the naming warnings. ✔
- Create a Unit Test project for the SecretSanta.Api project. ✔
  - Test all controller public methods. It is up to you to determine the best way to provide the controller's dependencies. ❌
  - Ensure the project is named with a ".Tests" suffix ✔



This looks like a great start.  I don't see a unit test project for the business project which is why all of those requirements have an 'x'. And I left an 'x' for testing all the public methods in the api test project because I only see unit tests for the get method. 

Maybe I missed something, but it looks great overall so far!

File: SecretSanta/src/SecretSanta.Api/Dto/UpdateUser.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+namespace SecretSanta.Api.Dto
+{
+    public class UpdateUser{
```
Comment (Keboo) [inline SecretSanta/src/SecretSanta.Api/Dto/UpdateUser.cs:3]: Nit: Typically we put braces on their own line
```suggestion
    public class UpdateUser
    {

```

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/UsersControllerTests.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+namespace SecretSanta.Api.Tests.Controllers
+{
+    [TestClass]
+    public class UsersControllerTests
+    {
+        [TestMethod]
+        //[ExpectedException(typeof(ArgumentNullException))]
+        public void Constructor_WithNullUserManager_ThrowsAppropriateException()
+        {
+            //Any of the approachs shown here are fine.
```
Comment (Keboo) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/UsersControllerTests.cs:19]: Typically you would not use all three approaches, you would pick one and remove the others.

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/UsersControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                return;
+            }
+            Assert.Fail("No exception thrown");
+        }
+
+        [TestMethod]
+        public void Get_WithData_ReturnsUsers()
+        {
+            //Arrange
+            UsersController controller = new(new UserManager());
```
Comment (Keboo) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/UsersControllerTests.cs:40]:
```suggestion
            UsersController controller = new(new TestableUserManager());

```

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/UsersControllerTests.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+using System.Collections.Generic;
+using SecretSanta.Business;
+using SecretSanta.Data;
+using Microsoft.AspNetCore.Mvc;
+using System;
+
+namespace SecretSanta.Api.Tests.Controllers
+{
+    [TestClass]
+    public class UsersControllerTests
```
Comment (Keboo) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/UsersControllerTests.cs:13]: The tests that are here are good, but it appears that several of the other API methods are not tested.

**Review by Keboo — COMMENTED** [review]
What is here is good. Looks like the tests for the Business project were missed along with the rest of the tests for the API controller.

- Create a UserController in the API projects
  - Implement endpoints for:
    - Retrieving all users ✔
    - Retriving a single user by id ✔
    - Creating a user ✔
    - Updating a user ✔
    - Deleting a user ✔
  - The UserController should function. ✔
  - Include a `.rest` file for interacting with the controller. ✔
- The `IUserRepository` interface is provided for you. You will need to provide an implementation of this interface within the SecretSanta.Business project. This implementation can simply store the user within a list, dictionary, array, or similar. ✔
- All test projects should be within a test directory ✔
- Create a unit test project for the SecretSanta.Business project. ❌
  - Unit test all public methods ❌
  - Ensure the project is named with a ".Tests" suffix ❌
  - Ensure test class is named appropriately for the class being tested: ❌
  - Name all tests in the form `<MethodBeingTested>_<ConditionBeingTested>_<ExpectedOutcome>()`. You can add  `<NoWarn>CA1707;INTL0003</NoWarn>` to remove the naming warnings. ✔
- Create a Unit Test project for the SecretSanta.Api project. ✔
  - Test all controller public methods. It is up to you to determine the best way to provide the controller's dependencies. ❌
  - Ensure the project is named with a ".Tests" suffix ✔

**Pull request closed without merging**